tokio/src/runtime/task/list.rs

```diff
@@ -140,34 +146,33 @@ impl<S: 'static> OwnedTasks<S> {
 
     /// Shuts down all tasks in the collection. This call also closes the
     /// collection, preventing new items from being added.
-    pub(crate) fn close_and_shutdown_all(&self)
+    /// The parameter start should be random among different worker threads
+    /// to reduce lock conflicts during shutdown.
+    /// Initiate shutting down the segment indexed by the start, and reset to 0
+    /// once the segment_size is reached, continuing until start - 1, it works like a ring.
+    pub(crate) fn close_and_shutdown_all(&self, start: usize)
     where
         S: Schedule,
     {
-        // The first iteration of the loop was unrolled so it can set the
-        // closed bool.
-        let first_task = {
-            let mut lock = self.inner.lock();
-            lock.closed = true;
-            lock.list.pop_back()
-        };
-        match first_task {
-            Some(task) => task.shutdown(),
-            None => return,
-        }
-
-        loop {
-            let task = match self.inner.lock().list.pop_back() {
-                Some(task) => task,
-                None => return,
-            };
-
-            task.shutdown();
+        self.closed.store(true, Ordering::Release);
+        for i in start..self.segment_size as usize + start {
+            loop {
+                let mut lock = self.segment_inner(i);
+                let task = match lock.pop_back() {
+                    Some(task) => {
+                        drop(lock);
+                        self.count.fetch_sub(1, Ordering::Relaxed);
+                        task
+                    }
+                    None => break,
+                };
+                task.shutdown();
+            }
```

Comment:
style nit: i think we can simplify this to
```suggestion
            while let Some(task) = self.segment_inner(i).pop_back() {
                self.count.fetch_sub(1, Ordering::Relaxed);
                task.shutdown();
            }
```
since the `lock` is dropped at the end of the expression `self.segment_inner(i).pop_back()`. of course, if you think it improves readability to make the duration of the mutex lock more explicit, we could also write
```suggestion
            loop {
                let mut lock = self.segment_inner(i);
                match lock.pop_back() {
                    Some(task) => {
                        drop(lock);
                        self.count.fetch_sub(1, Ordering::Relaxed);
                        task.shutdown();
                    }
                    None => break,
                };
            }
```
but IMO, the `while` loop is a bit nicer. it's up to you, though.

Reply:
When using `while let Some(task) = self.segment_inner(i).pop_back() {`, we encounter the undesirable effect of holding the mutex for a longer time. The MutexGuard won't be released until the end of an iteration of the while loop.

The `task.shutdown` operation should be performed when we do not hold the lock.

Reply:
I would like to take the second suggestion, but I made the mistake of taking the first suggestion so I chose to submit the code myself instead of adding suggestion to batch.

Reply:
> When using while let Some(task) = self.segment_inner(i).pop_back() {, we encounter the undesirable effect of holding the mutex for a longer time. The MutexGuard won't be released until the end of an iteration of the while loop.

hmm, i thought that the MutexGuard would be dropped at the end of the `self.segment_inner(i).pop_back()` expression, so the mutex is locked for the same period of time as in the current code. but, if that's not actually the case, i agree that the current code is better, since the critical section is only as long as it takes to pop the task.

would introducing an explicit scope in the `while` loop also have that effect? like this:
```rust
while let Some(task) = { self.segment_inner(i).pop_back() } {
    // ...
}
```
all locals to the scope (like the guard) should be dropped as soon as the scope is evaluated, right?

Reply:
in any case, i don't have a problem with the current code as-is, i'm just curious about whether the while loop would have similar behavior or not.

Reply:
The lifetime of `MutexGuard` is as follows:
```rust
while let Some(task) = self.segment_inner(i).pop_back() { // get MutexGuard here
      self.count.fetch_sub(1, Ordering::Relaxed);
      let _lock = self.segment_inner(i); // dead lock here
      task.shutdown();
}// drop MutexGuard here
```
The dead lock line reflects the lifetime of `MutexGuard`.

I belive the auto drop feature of Rust is very useful for avoid unexpected resources not being released, but in order to achieve the precise lock control, the better way is that we drop lock manully like other languages which has no auto drop.

Reply:
The following does not change the lifetime of `MutexGuard` at all.
```
while let Some(task) = { self.segment_inner(i).pop_back() } {
    // ...
}
```
- we create `MutexGuard` in ` { self.segment_inner(i).pop_back() } ` code block
- and then move it from the block
- the `MutexGuard` still will not be dropped until th end of an iteration of this while loop.

This behavior of `MutexGuard in while let` also can be found in https://stackoverflow.com/questions/58968488/why-is-this-mutexguard-not-dropped